src/agent/thread_helper.cpp

```diff
@@ -0,0 +1,361 @@
+    SuccessOrExit(err = otLinkSetPanId(mInstance, aPanId));
+    SuccessOrExit(err = otThreadSetExtendedPanId(mInstance, &extPanId));
+    SuccessOrExit(err = otThreadSetMasterKey(mInstance, &masterKey));
+
+    channelMask = otPlatRadioGetPreferredChannelMask(mInstance) & aChannelMask;
+
+    if (channelMask == 0)
+    {
+        channelMask = otLinkGetSupportedChannelMask(mInstance) & aChannelMask;
+    }
+    VerifyOrExit(channelMask != 0, err = OT_ERROR_INVALID_ARGS);
+
+    channel = RandomChannelFromChannelMask(channelMask);
+    SuccessOrExit(otLinkSetChannel(mInstance, channel));
+
+    pskcRet = otThreadSetPskc(mInstance, &pskc);
+    VerifyOrExit(pskcRet == OT_ERROR_NONE, err = OT_ERROR_FAILED);
+
+    if (!otIp6IsEnabled(mInstance))
+    {
+        SuccessOrExit(err = otIp6SetEnabled(mInstance, true));
+    }
+    SuccessOrExit(err = otThreadSetEnabled(mInstance, true));
+exit:
+    if (err != OT_ERROR_NONE)
+    {
+        if (aHandler)
+        {
+            aHandler(err);
+        }
+        mAttachHandler = nullptr;
+    }
+}
+
+otError ThreadHelper::Reset(void)
+{
+    mDeviceRoleHandlers.clear();
+    otInstanceReset(mInstance);
+
+    return OT_ERROR_NONE;
+}
+
+void ThreadHelper::JoinerStart(const std::string &aPskd,
+                               const std::string &aProvisioningUrl,
+                               const std::string &aVendorName,
+                               const std::string &aVendorModel,
+                               const std::string &aVendorSwVersion,
+                               const std::string &aVendorData,
+                               ResultHandler      aHandler)
+{
+    otError err = OT_ERROR_NONE;
+
+    VerifyOrExit(aHandler != nullptr, err = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(mAttachHandler == nullptr && mJoinerHandler == nullptr, err = OT_ERROR_INVALID_STATE);
+    mJoinerHandler = aHandler;
+
+    if (!otIp6IsEnabled(mInstance))
+    {
+        SuccessOrExit(err = otIp6SetEnabled(mInstance, true));
+    }
+    err = otJoinerStart(mInstance, aPskd.c_str(), aProvisioningUrl.c_str(), aVendorName.c_str(), aVendorModel.c_str(),
+                        aVendorSwVersion.c_str(), aVendorData.c_str(), sJoinerCallback, this);
+exit:
+    if (err != OT_ERROR_NONE)
+    {
+        if (aHandler)
+        {
+            aHandler(err);
+        }
+        mJoinerHandler = nullptr;
+    }
+}
+
+void ThreadHelper::sJoinerCallback(otError aError, void *aThreadHelper)
+{
+    ThreadHelper *helper = static_cast<ThreadHelper *>(aThreadHelper);
+
+    helper->JoinerCallback(aError);
+}
+
+void ThreadHelper::JoinerCallback(otError aError)
+{
+    if (aError != OT_ERROR_NONE)
+    {
+        otIp6SetEnabled(mInstance, false);
+        mJoinerHandler(aError);
+    }
+    else
+    {
+        otThreadSetEnabled(mInstance, true);
+    }
+    mJoinerHandler = nullptr;
+}
+
+otError ThreadHelper::AddUnsecurePort(uint16_t aPort, uint32_t aSeconds)
```

Comment:
Yes, it is. There is not a mechanism to remove this port since nobody is using the remove unsecure port functionality.
I'll add some checks for adding the same port.

Reply:
I know next layer do use `wpantund` "permit-join" with 0 as timeout to remove the port (make device non-joinable).